_posts/2020-07-27-creating-a-snap-chat-mini.markdown

```diff
@@ -0,0 +1,194 @@
+
+    // restore cursor position
+    e.target.setSelectionRange(cursorPosition, cursorPosition)
+  }
+}
+```
+
+If you use an element that's not an input with the `contenteditable` attribute, you may need to do some changes to that snippet in order to get the current cursor position from the document element.
+
+### Changing Input Case Breaks iOS Keyboard Suggestions
+
+One initial requirement we had for an input element was that it should always by uppercase characters. The easiest solution for this requirement would be to use a CSS text transform property:
+
+```css
+#my-input {
+  text-transform: uppercase;
+}
+```
+
+But this creates an issue. This modifies the element's value and iOS keyboard loses track of the original word it has to replace if you select one of the suggestions, not being able to actually apply the suggestion.
+
+We also tried a different solution for that requirement listening to the `input` event and upcasing the letters after the input changed, but this creates the same issue for iOS' keyboard suggestions.
+
+One solution we didn't try (because it would be too hacky and we decide to remove the requirement) is to have 2 different inputs, one where the user will actually type (hidden somehow using `visibility: hidden`) and one that shows the original input value but upcased.
+
+I would advice not to have that requirement unless it's critical to your app or you don't mind having the keyboard suggestions functionality broken on that input for iOS.
+
+### DONE Key on iOS is not a Key
+
+iOS keyboard shows a `DONE` "key" as a part of the keyboard, but this is does not work like a key, you can't detect it being pressed (iOS keyboard does trigger keypress events) and it won't work like the `return` or `enter` key either submitting a form. When pressing this key, it will just blur/unfocus the current element and hide the keyboard.
+
+My advice here is that, if you have a form, you should have an specific `submit` element. If you relay on the user pressing `enter` or `return` keys, that might be counterintuitive for iOS users used to press `DONE` when they finished typing.
+
+### Repaint Issues
+
+When you are working with animations, the browsers will try to optimize the repaints so improve performance. We encountered some really specific cases where, for example, an animation happened too quick outside the viewport and it had the wrong final state.
+
+The solution we had here was to force a repaint of the element at the time we detected it was inside the viewport:
+
+```javascript
+afterEnteringTheViewport(element) {
+  element.style.transform = 'translateZ(0)'
+}
+```
+
+> This is just an example of a function, it's not a real callback function you can use like an event handler. How you detect that the element enters the viewport will depend on your app's implementation. You could use the `IntersectionObserver` for example or some css animation-related events.
+
+### Safari can't Handle `box-shadow` and a Gradient Background
```

Comment:
Hmmm, I'm not sure that this is specific to snap minis and therefor if it belongs here?